src/main/kotlin/frc/robot/subsystems/SwerveModule.kt

```diff
@@ -0,0 +1,183 @@
+import edu.wpi.first.math.kinematics.SwerveModuleState
+import edu.wpi.first.math.trajectory.TrapezoidProfile
+import edu.wpi.first.math.util.Units
+import edu.wpi.first.wpilibj.Timer
+import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard.getTab
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
+import edu.wpi.first.wpilibj2.command.SubsystemBase
+import frc.robot.Constants.DRIVE_GEAR_RATIO
+import frc.robot.Constants.WHEEL_CIRCUMFRENCE
+import frc.robot.controls.ControlScheme
+
+class SwerveModule(
+    driveMotorId: Int,
+    turnMotorId: Int,
+    encoderId: Int,
+    val mname: String,
+    val position: Translation2d,
+    angleZero: Double = 0.0,
+    val controlScheme: ControlScheme
+) : SubsystemBase() {
+    val stateEntry = getTab("Swerve Diagnostics")
+        .add("$mname State v", 0)
+        .entry
+    val driveMotor = WPI_TalonFX(driveMotorId).apply {
+        configFactoryDefault()
+        configSupplyCurrentLimit(SupplyCurrentLimitConfiguration(true, 40.0, 45.0, 0.0)) // why 0.5?
+        //driveMotor.configClosedloopRamp(0.25);
+        configStatorCurrentLimit(StatorCurrentLimitConfiguration(true, 40.0, 45.0, 0.5))
+        setNeutralMode(NeutralMode.Brake)
+        // Configure the encoders for both motors
+//        configSelectedFeedbackSensor(TalonFXFeedbackDevice.IntegratedSensor, 0, 0)
+    }
+    private val turnMotor = WPI_TalonFX(turnMotorId).apply {
+        configFactoryDefault()
+//        setStatusFramePeriod(StatusFrameEnhanced.Status_3_Quadrature, 255)
+//        setStatusFramePeriod(StatusFrameEnhanced.Status_2_Feedback0, 255)
+//        setStatusFramePeriod(StatusFrameEnhanced.Status_1_General, 255)
+        setNeutralMode(NeutralMode.Brake)
+    }
+    private val turnEncoder = CANCoder(encoderId).apply {
+        configAbsoluteSensorRange(AbsoluteSensorRange.Signed_PlusMinus180)
+        configMagnetOffset(-1 * angleZero)
+//        setStatusFramePeriod(CANCoderStatusFrame.SensorData, 10, 100)
+
+        configSensorDirection(false)
+
+    }
+
+    // smartdashboard
+    // motors
+    val tab = getTab("$mname Swerve Diagnostics")
+    val driveMotorEntry = tab.add("$mname Drive Motor", 0.0).entry
+    val turnMotorEntry = tab.add("$mname Turn Motor", 0.0).entry
+    val turnEncoderEntry = tab.add("$mname Turn Encoder", 0.0).entry
+
+    // motor sets
+    val driveMotorSetEntry = tab.add("$mname Drive Motor Set", 0.0).entry
+    val turnMotorSetEntry = tab.add("$mname Turn Motor Set", 0.0).entry
+
+    // position/angle
+    val positionEntry = tab.add("$mname Position", 0.0).entry
+    val angleEntry = tab.add("$mname Angle", 0.0).entry
+
+    val DRIVE_P = 2.37
+    val DRIVE_I = 0.0
+    val DRIVE_D = 0.0
+    val drivePid = ProfiledPIDController(
+        DRIVE_P,
+        DRIVE_I,
+        DRIVE_D,
+        TrapezoidProfile.Constraints(2.0, .0)// TODO: Fix these
+    )
+    private val driveFF = SimpleMotorFeedforward(0.21862, 2.2997, 0.26242)
+
+
+    val ANGLE_P = 0.5
```

Comment:
This P value might be too low. I think we should double-check the sysid value and make sure it's in position mode